### PR TITLE
 [Q3] Auto-update asset list at regular intervals

### PR DESCRIPTION
**Pull request by deividaspetraitis**
This PR implements updating asset list at regular intervals based on the procesing block in the ingester, please see #279 for more information.

## Implementation details

- As a side effect several more mocks were created in order to be able to test ingester use case, and existing mock was modified to support future testing requirements.

----

Old pull request: https://github.com/osmosis-labs/sqs/pull/323

## Summary by CodeRabbit

- **New Features**
  - Automatic asset list updates at specified intervals to keep your asset information up-to-date.

- **Bug Fixes**
  - Enhanced asset update process to ensure smooth and non-blocking block processing.

- **Documentation**
  - Updated API documentation with new configuration parameters for asset updates and token metadata endpoint.

- **Tests**
  - Added new test functions to validate asset list updates and chain denomination.

- **Chores**
  - Incorporated a no-op logger for improved test logging consistency.

### CODE REVIEW

**Comment by deividaspetraitis** [comment]
Things not addressed because of more info neede in the last pull request, cc @p0mvn:

- ~~https://github.com/osmosis-labs/sqs/pull/323#discussion_r1666544719~~
- ~~https://github.com/osmosis-labs/sqs/pull/323#discussion_r1664579527~~
- ~~https://github.com/osmosis-labs/sqs/pull/323#discussion_r1665920048~~
- ~~https://github.com/osmosis-labs/sqs/pull/323#discussion_r1666577205~~

File: tokens/usecase/tokens_usecase.go
Comment (deividaspetraitis) [inline tokens/usecase/tokens_usecase.go:51]: Logger is no longer used at this moment, but I think it's still useful to have it. Please let me know if I should remove it.

File: tokens/usecase/tokens_usecase.go
Comment (deividaspetraitis) [inline tokens/usecase/tokens_usecase.go]: Instead of returning `mvc.TokensUsecase`, return `*tokensUseCase`

File: tokens/usecase/token_registry_test.go
Comment (deividaspetraitis) [inline tokens/usecase/token_registry_test.go]: Move from package `usecase` to `usecase_test`

File: tokens/usecase/token_registry.go
Comment (deividaspetraitis) [inline tokens/usecase/token_registry.go]: Move `loadTokens LoadTokensFunc` to `ChainRegistryHTTPFetcher`, and add method like `tokensusecase.SetTokenRegistryLoader(TokenRegistryLoader)`

File: ingest/usecase/ingest_usecase.go
Comment (p0mvn) [inline ingest/usecase/ingest_usecase.go]: Let's consolidate all metrics in a unified place here please:
https://github.com/osmosis-labs/sqs/blob/v25.x/domain/telemetry.go

Would appreciate following the same format as in that file 🙏

File: ingest/usecase/ingest_usecase.go
Comment (p0mvn) [inline ingest/usecase/ingest_usecase.go]: Consider creating a method defined on the ingester usecase called `UpdateAssetsAtHeightIntervalAsync` that abstracts this.

Would like to avoid spinning off many goroutines in this execution flow for readability, thus, instead creating these "async" wrappers so that it is easier to follow

File: tokens/usecase/tokens_usecase_test.go
Comment (p0mvn) [inline tokens/usecase/tokens_usecase_test.go]: For this and other tests, let's please move the errors to `domain/errors.go` so that if we change an error message, we only need to do it in one place.

Additionally, sometimes we might require reusing the same error when testing some of the upstream methods.

See an example [here](https://github.com/osmosis-labs/sqs/blob/48713b07f622e2c377d37598ab77f4b6d7e18ede/router/usecase/optimized_routes_test.go#L345)

File: tokens/usecase/tokens_usecase_test.go
Comment (p0mvn) [inline tokens/usecase/tokens_usecase_test.go]: Does it make sense moving this into the `domain/mocks` package?

File: tokens/usecase/tokens_usecase_test.go
Comment (p0mvn) [inline tokens/usecase/tokens_usecase_test.go]: Let's please use the following pattern for stylistic consistency and readability
```

if tt.expectedErr != nil {
   s.Require().Error(err)
   s.Require().ErrorIs(tt.expectedErr, err)
   return 
}

s.Require().NoError(err)

...
```

File: domain/mocks/candidate_route_search_data_worker_mock.go
Comment (p0mvn) [inline domain/mocks/candidate_route_search_data_worker_mock.go:14]: I do not see this being used anywhere. If still unused, why are we adding it?

File: tokens/usecase/tokens_usecase.go
Comment (p0mvn) [inline tokens/usecase/tokens_usecase.go]: IMO we can remove this check

If the app wiring issue occurs, I would rather have it panic early due to nil pointer dereferencing rather than battling a prod issue debugging why the asset list doesn't update and there is no error

**Review by p0mvn — COMMENTED** [review]
Nice work!

The remaining items are mostly nits.

I have one question/request before doing a final approval:
- Have we tested this with an integration test suite against a mainnet node?

That is, run a mainnet node + sqs instance and then do:

```
make e2e-run-local
```

If there are any failures, please link them here via comment 🙏

File: ingest/usecase/ingest_usecase.go
Comment (deividaspetraitis) [reply ingest/usecase/ingest_usecase.go]: I have updated PR to reflect this suggestion, thanks! I must looked at the wrong [example](https://github.com/osmosis-labs/sqs/blob/d9f566eb739d4a499a6c86ba66e8848997ec9801/router/usecase/router_usecase.go#L58), should we create a task refactoring such places too?

File: tokens/usecase/tokens_usecase_test.go
Comment (deividaspetraitis) [reply tokens/usecase/tokens_usecase_test.go]: I have created distinct error types as per [your example](https://github.com/osmosis-labs/sqs/blob/48713b07f622e2c377d37598ab77f4b6d7e18ede/router/usecase/optimized_routes_test.go#L345), however instead of moving to `domain/errors.go` I left them in  `tokens/usecase/errors.go`,  since they seems too verbose for domain at this point, please let me know if I should move to `domain/errors.go`. In general it seems that number of errors started growing and my concern is that we can end up exposing too much implementation details and making API less flexible, IMHO it would be useful to came up with more general error messages, at this point I wanted to avoid refactoring actual messages in case to clients relies on them.

File: tokens/usecase/tokens_usecase_test.go
Comment (deividaspetraitis) [reply tokens/usecase/tokens_usecase_test.go]: Moved to `domain/mocks`. :)

File: ingest/usecase/ingest_usecase.go
Comment (deividaspetraitis) [reply ingest/usecase/ingest_usecase.go]: Moved to `updateAssetsAtHeightIntervalAsync` method. Note that it's unexported to keep API clean and not reveal implementation details, let me know if you think it should be exported.

File: tokens/usecase/tokens_usecase_test.go
Comment (deividaspetraitis) [reply tokens/usecase/tokens_usecase_test.go]: Good catch, updated, thanks!

File: tokens/usecase/tokens_usecase.go
Comment (deividaspetraitis) [reply tokens/usecase/tokens_usecase.go]: Updated as per your suggestion.

File: domain/mocks/candidate_route_search_data_worker_mock.go
Comment (deividaspetraitis) [reply domain/mocks/candidate_route_search_data_worker_mock.go:14]: This one is used in the [tests here](https://github.com/osmosis-labs/sqs/blob/d9f566eb739d4a499a6c86ba66e8848997ec9801/ingest/usecase/ingest_usecase_test.go#L437)

File: tokens/usecase/tokens_usecase_test.go
Comment (deividaspetraitis) [reply tokens/usecase/tokens_usecase_test.go]: Discussed with Roman, we should also add some more format documentation in CONTRIBUTING.md. ( Create a task ).

File: ingest/usecase/ingest_usecase.go
Comment (deividaspetraitis) [reply ingest/usecase/ingest_usecase.go]: TODO: for now create task to move other metrics to https://github.com/osmosis-labs/sqs/blob/v25.x/domain/telemetry.go

**Comment by cryptomatictrader** [comment]
Sorry for being late for catching up on this, and thanks for addressing my previous comments. Nice job @deividaspetraitis

File: ingest/usecase/ingest_usecase.go
Comment (deividaspetraitis) [reply ingest/usecase/ingest_usecase.go]: Task [DATA-269 ](https://linear.app/osmosis/issue/DATA-269/consolidate-metrics-in-metricsgo) was created to address this issue.

File: ingest/usecase/ingest_usecase.go
Comment (deividaspetraitis) [reply ingest/usecase/ingest_usecase.go]: As per our discussion this is expected approach, marking as resolved.

File: tokens/usecase/tokens_usecase_test.go
Comment (deividaspetraitis) [reply tokens/usecase/tokens_usecase_test.go]: [DATA-270](https://linear.app/osmosis/issue/DATA-270/refactor-and-or-formalise-errors-in-contributingmd-document) was created.

**Comment by deividaspetraitis** [comment]
> Nice work!
> 
> The remaining items are mostly nits.
> 
> I have one question/request before doing a final approval:
> 
> * Have we tested this with an integration test suite against a mainnet node?
> 
> That is, run a mainnet node + sqs instance and then do:
> 
> ```
> make e2e-run-local
> ```
> 
> If there are any failures, please link them here via comment 🙏

Hey @p0mvn, I have run e2e tests against mainnet node as per you suggestion, following are the result. It seems that several tests are failing, I was wondering how I can fix that, any suggestions? Full output can be found in [e2e-run-local.txt](https://github.com/user-attachments/files/16159655/e2e-run-local.txt).

```
tests/test_pools.py:61: AssertionError
======================================================================================================= short test summary info ========================================================================================================
FAILED tests/test_pools.py::TestPools::test_pools_pool_liquidity_cap[http://localhost:9092-722] - AssertionError: ID (722) Pool liquidity cap was 264918 - expected 239864.5965028432, actual error 0.09457040856852612 error tolerance 0.05
FAILED tests/test_pools.py::TestPools::test_pools_pool_liquidity_cap[http://localhost:9092-1423] - AssertionError: ID (1423) Pool liquidity cap was 94441 - expected 101442.58760091313, actual error 0.06902019917372561 error tolerance 0.05
FAILED tests/test_pools.py::TestPools::test_pools_pool_liquidity_cap[http://localhost:9092-1816] - AssertionError: ID (1816) Pool liquidity cap was 1382860 - expected 1005423.8299677123, actual error 0.272938815232408 error tolerance 0.05
============================================================================================== 3 failed, 1470 passed, 6 skipped in 8.34s ===============================================================================================
```

**Comment by p0mvn** [comment]
> > Nice work!
> > The remaining items are mostly nits.
> > I have one question/request before doing a final approval:
> > 
> > * Have we tested this with an integration test suite against a mainnet node?
> > 
> > That is, run a mainnet node + sqs instance and then do:
> > ```
> > make e2e-run-local
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If there are any failures, please link them here via comment 🙏
> 
> Hey @p0mvn, I have run e2e tests against mainnet node as per you suggestion, following are the result. It seems that several tests are failing, I was wondering how I can fix that, any suggestions? Full output can be found in [e2e-run-local.txt](https://github.com/user-attachments/files/16159655/e2e-run-local.txt).
> 
> ```
> tests/test_pools.py:61: AssertionError
> ======================================================================================================= short test summary info ========================================================================================================
> FAILED tests/test_pools.py::TestPools::test_pools_pool_liquidity_cap[http://localhost:9092-722] - AssertionError: ID (722) Pool liquidity cap was 264918 - expected 239864.5965028432, actual error 0.09457040856852612 error tolerance 0.05
> FAILED tests/test_pools.py::TestPools::test_pools_pool_liquidity_cap[http://localhost:9092-1423] - AssertionError: ID (1423) Pool liquidity cap was 94441 - expected 101442.58760091313, actual error 0.06902019917372561 error tolerance 0.05
> FAILED tests/test_pools.py::TestPools::test_pools_pool_liquidity_cap[http://localhost:9092-1816] - AssertionError: ID (1816) Pool liquidity cap was 1382860 - expected 1005423.8299677123, actual error 0.272938815232408 error tolerance 0.05
> ============================================================================================== 3 failed, 1470 passed, 6 skipped in 8.34s ===============================================================================================
> ```

Thanks for the summary. These are getting fixed by concurrent PRs - we can ignore them here

File: ingest/usecase/ingest_usecase.go
Comment (p0mvn) [inline ingest/usecase/ingest_usecase.go:180]: Create an alert for domain.SQSUpdateAssetsAtHeightIntervalErrorCounter

**Review by p0mvn — APPROVED** [review]
Nice work!

**Pull request merged**
